src/DataStructures/Tensor/EagerMath/Magnitude.hpp

```diff
@@ -48,11 +50,56 @@ Scalar<DataType> magnitude(
 /// \details
 /// Computes the square root of the absolute value of the scalar.
 template <typename DataType>
-Scalar<DataType> sqrt_magnitude(
-    const Scalar<DataType>& input) noexcept {
+Scalar<DataType> sqrt_magnitude(const Scalar<DataType>& input) noexcept {
   return Scalar<DataType>{sqrt(abs(get(input)))};
 }
 
+/// \ingroup TensorGroup
+/// \brief Compute point-wise $\\ell^2$ (Frobenius) norm of a tensor
+///
+/// \details
+/// Computes the point-wise Frobenius norm of a given tensor of arbitrary rank.
+template <typename DataType, typename Symm, typename IndexList>
+Scalar<DataType> pointwise_l2_norm(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  DataType pointwise_l2_normsq{tensor.begin()->size(), 0.};
+  for (auto& tensor_element : tensor) {
+    // In order to handle tensory symmetries, we multiply the square of each
+    // stored component with its multiplicity
+    pointwise_l2_normsq +=
+        tensor.multiplicity(const_cast<DataType*>(&tensor_element)) *  // NOLINT
+        square(tensor_element);
+    // `const_cast` above is necessary to stop gcc warnings (errors) about
+    // the implicit conversion of function argument from a pointer to const to
+    // pointer to non-const type.
+  }
+  return Scalar<DataType>{sqrt(pointwise_l2_normsq)};
+}
+
+/// \ingroup TensorGroup
+/// \brief Compute $\\ell^2$ (Frobenius) norm of a tensor
+///
+/// \details
+/// Computes the Frobenius norm of a given tensor of arbitrary rank,
+/// averaged over all grid points.
+template <typename DataType, typename Symm, typename IndexList>
+double l2_norm(const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  DataType pointwise_l2_normsq{tensor.begin()->size(), 0.};
+  for (auto& tensor_element : tensor) {
+    // In order to handle tensory symmetries, we multiply the square of each
+    // stored component with its multiplicity
+    pointwise_l2_normsq +=
+        tensor.multiplicity(const_cast<DataType*>(&tensor_element)) *  // NOLINT
+        square(tensor_element);
+    // The `const_cast` above is necessary to stop gcc warning (error) about
+    // the implicit conversion of function argument from a pointer to const to
+    // pointer to non-const type.
```

Comment:
Thanks for this comment, but I still don't quite understand why the `const_cast` is needed. The `multiplicity` function takes an iterator by const-reference. I don't know if what you are doing here and the `NOLINT` is the right thing to do. Perhaps someone from @sxs-collaboration/spectre-core-devs can have a quick look?

Reply:
You've found a bug in `Tensor`.  Apply this patch:
```patch
diff --git a/src/DataStructures/Tensor/Tensor.hpp b/src/DataStructures/Tensor/Tensor.hpp
index 31882e4f8..f5d3aa71e 100644
--- a/src/DataStructures/Tensor/Tensor.hpp
+++ b/src/DataStructures/Tensor/Tensor.hpp
@@ -353,7 +353,7 @@ class Tensor<X, Symm, IndexList<Indices...>> {
   ///
   /// \see TensorMetafunctions::compute_multiplicity
   SPECTRE_ALWAYS_INLINE constexpr size_t multiplicity(
-      const iterator& iter) const noexcept {
+      const const_iterator& iter) const noexcept {
     return structure::multiplicity(static_cast<size_t>(iter - begin()));
   }
   SPECTRE_ALWAYS_INLINE static constexpr size_t multiplicity(
diff --git a/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp b/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
index bf89c22b4..2e2c2efe5 100644
--- a/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
+++ b/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
@@ -1150,7 +1150,7 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Multiplicity",
       tensor;
 
   // Test multiplicity by iterator
-  for (auto it = tensor.begin(); it != tensor.end(); ++it) {
+  for (auto it = tensor.cbegin(); it != tensor.cend(); ++it) {
     const auto& indices = tensor.get_tensor_index(it);
     const size_t multiplicity = (indices[1] == indices[2] ? 1 : 2);
     CHECK(multiplicity == tensor.multiplicity(it));
```

However, I don't think there is any guarantee that the Tensor iterators are pointers, so passing `&tensor_element` is not guaranteed to work.  You should explicitly use the iterators rather than a range-based loop.  (You will still need the patch above.)

Reply:
> The `multiplicity` function takes an iterator by const-reference.

Yes `multiplicity` takes as input a `const reference` to an `iterator`. The `iterator` itself is pointer without a `const` qualifier. So, gcc deduces that `multiplicity` should take a `const reference` to a `non-const pointer`.  However, `&tensor_element` results in a `const pointer`, and the line above would try to pass to `multiplicity` a `const reference` to that `const pointer` (if i hadn't casted away the `const`). At least, this is what I understood from the error message i get when i don't `const_cast` in the line above:

```
/home/prayush/src/spectre/spectre_fix_rnd_gh_failures/src/DataStructures/Tensor/Tensor.hpp:355:42: note: candidate: constexpr size_t Tensor<X, Symm, IndexList<Indices ...> >::multiplicity(const iterator&) const [with X = DataVector; Symm = brigand::list<>; IndexList = brigand::list; Indices = {}; size_t = long unsigned int; Tensor<X, Symm, IndexList<Indices ...> >::iterator = DataVector*] <near match>
   SPECTRE_ALWAYS_INLINE constexpr size_t multiplicity(
                                          ^~~~~~~~~~~~
/home/prayush/src/spectre/spectre_fix_rnd_gh_failures/src/DataStructures/Tensor/Tensor.hpp:355:42: note:   conversion of argument 1 would be ill-formed:
In file included from /home/prayush/src/spectre/spectre_fix_rnd_gh_failures/tests/Unit/DataStructures/Tensor/EagerMath/Test_Magnitude.cpp:14:0:
/home/prayush/src/spectre/spectre_fix_rnd_gh_failures/src/DataStructures/Tensor/EagerMath/Magnitude.hpp:68:48: **error: invalid conversion from ‘const DataVector*’ to ‘Tensor<DataVector, brigand::list<>, brigand::list<> >::iterator {aka DataVector*}’ [-fpermissive]**
         tensor.multiplicity((&tensor_element)) *  // NOLINT
```


> I don't know if what you are doing here and the `NOLINT` is the right thing to do. Perhaps someone from @sxs-collaboration/spectre-core-devs can have a quick look?

This is one possible workaround. If there's a cleaner way to do this, i'd be happy to learn.

Reply:
Aha @wthrowe -- saw your message just now! Thanks for changing `multiplicity`.

Reply:
Applied will's patch to make `multiplicity` work with `const iterator`s, and removed the `const_cast` now.